### ISSUE TITLE
dotnet restore should restore DotNetCliToolReference items across P2Ps

### ISSUE THREAD

**Issue by AArnott** [NuGet/Home#5039]
As first discussed in dotnet/sdk#1123, we eventually found that a build failure was due to `dotnet restore` not restoring `DotNetCliToolReference` items across P2P references. So if I `dotnet restore test`, the build can fail because the referenced library project requires a `DotNetCliToolReference` to be restored in order to build. 

**Impact:** Both @ctaggart's SourceLink and my own CodeGeneration.Roslyn MSBuild extensions rely on dotnet CLI tools in their build because supporting both MSBuild full and MSBuild Core for interesting MSBuild tasks is *so danged hard*. 
This means that every build that uses either of these (or other similar) build extensions will have to take care to restore each individual project (or via the sln) to ensure the build doesn't fail.

Can `dotnet restore` be made to restore CLI tools across P2P boundaries as it already does for package references?

```
dotnet.exe --info

.NET Command Line Tools (1.0.1)

Product Information:
 Version:            1.0.1
 Commit SHA-1 hash:  005db40cd1

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  14.04
 OS Platform: Linux
 RID:         ubuntu.14.04-x64
 Base Path:   /usr/share/dotnet/sdk/1.0.1
```

**Comment by emgarten** [comment]
This is fixed in MSBuild full with VS 15.1, and in dotnet cli 2.0.0

https://github.com/NuGet/Home/issues/4711

**Comment by AArnott** [comment]
Thanks.

**Comment by verysimplenick** [comment]
This is not fixed in 2.0.2 and vs 15.3:
```
dotnet --info
Программы командной строки .NET (2.0.2-vspre-006949)

Product Information:
 Version:            2.0.2-vspre-006949
 Commit SHA-1 hash:  23fc6ec23d

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.0.2-vspre-006949\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0
  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d

```
Works on build one project, but if I build solution with references it's failing.

```
C:\MyProject\project>dotnet restore MyProject.Core
  Restore completed in 34,85 ms for C:\MyProject\project\MyProject.Core\MyProject.Core.csproj.
  Restore completed in 13,7 ms for C:\MyProject\project\MyProject.Core\MyProject.Core.csproj.

C:\MyProject\project>dotnet restore MyProject.sln
  Restoring packages for C:\MyProject\project\MyProject.Tests\MyProject.Tests.csproj...
  Restoring packages for C:\MyProject\project\MyProject\MyProject.csproj...
  Restore completed in 27,52 ms for C:\MyProject\project\MyProject\MyProject.csproj.
  Restore completed in 37,94 ms for C:\MyProject\project\MyProject.Core\MyProject.Core.csproj.
  Restore completed in 32,26 ms for C:\MyProject\project\MyProject.Core\MyProject.Core.csproj.
  Restore completed in 45,53 ms for C:\MyProject\project\MyProject\MyProject.csproj.
C:\MyProject\project\MyProject.Tests\MyProject.Tests.csproj : error NU1105: Unable to find project information for 'C:\MyProject\project\MyProject.Core\MyProject.Core.csproj'. The project file may be invalid or missing targets required for restore. [C:\MyProject\project\MyProject.sln]
C:\MyProject\project\MyProject\MyProject.csproj : error NU1105: Unable to find project information for 'C:\MyProject\project\MyProject.Core\MyProject.Core.csproj'. The project file may be invalid or missing targets required for restore. [C:\MyProject\project\MyProject.sln]
  Restore failed in 282,6 ms for C:\MyProject\project\MyProject.Tests\MyProject.Tests.csproj.
  Restore failed in 410,86 ms for C:\MyProject\project\MyProject\MyProject.csproj.
```

If in `.Core` project I comment this line:

```
<DotNetCliToolReference Include="Microsoft.EntityFrameworkCore.Tools.DotNet" Version="2.0.0" PrivateAssets="All" />
```

Everything builds fine, but I don't have `dotnet ef` command :(

Maybe linked with another bug in telemetry [off]
```
#####>dotnet restore -v --no-cache
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRange_IndexException()
   at System.SZArrayHelper.get_Item[T](Int32 index)
   at System.Linq.Enumerable.ElementAt[TSource](IEnumerable`1 source, Int32 index)
   at Microsoft.DotNet.Cli.Telemetry.TelemetryFilter.LogVerbosityForAllTopLevelCommand(ICollection`1 result, ParseResult parseResult, String topLevelCommandName)
   at Microsoft.DotNet.Cli.Telemetry.TelemetryFilter.Filter(Object objectToFilter)
   at Microsoft.DotNet.Cli.Utils.TelemetryEventEntry.SendFiltered(Object o)
   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)
   at Microsoft.DotNet.Cli.Program.Main(String[] args)
```

**Comment by emgarten** [comment]
@verysimplenick I think you are hitting other issues here. The original problem is that tools in projects not directly being restored were skipped. If any at point the tool has been restored and it exists in `%UserProfile%/.nuget/packages/.tools` then it should work and may not be a restore issue.

Would you open a new issue to track this and include a simple solution that repros this? It can be either a github repro or zip file.
